### DISCUSSION TITLE
Clean Architecture - An Introduction | Dan Does Code

### DISCUSSION THREAD

**Discussion by giscus-migrator** [danielmackay/nextjs-personal-blog#64] [Blog Comments]
Clean Architecture - An Introduction
        
[https://www.dandoescode.com/blog/clean-architecture-an-introduction/](https://www.dandoescode.com/blog/clean-architecture-an-introduction/)

**Comment by giscus-migrator** [comment]
```header
Originally posted by Dave Mateer on 2020-05-08 17:07
```

Good article Dan. I like your conclusion about software becoming more testable and maintainable using patterns like this as that is where I spend most of my time (legacy code!)

**Reply by giscus-migrator** [reply]
```header
Originally posted on 2020-05-10 07:12
```

Cheers Dave! I'm really digging clean architecture with CQRS and MediatR. I have a new project starting where I will be using this pattern in full force. In projects I've previously worked on that use N-Tier, the 'services' end up becoming dumping grounds for all sorts of logic. I'm hoping CQRS and clean separation will help to solve this. Psyched to see the result.

**Reply by giscus-migrator** [reply]
```header
Originally posted by Dave Mateer on 2020-05-14 18:24
```

[https://app.pluralsight.com...](https://app.pluralsight.com/library/courses/ddd-ef-core-preserving-encapsulation/discussion) Vladimir Khorikov's courses on Pluralsight seem to be architecting along the same lines as you are, and this new one using EF Core and CQRS looks interesting too.

**Reply by giscus-migrator** [reply]
```header
Originally posted on 2020-05-17 12:28
```

I did a couple of pluralsight courses during free april around DDD and Clean Architecture. Vladimir's courses were on my list also, but ran out of time.

**Reply by giscus-migrator** [reply]
```header
Originally posted by Dave Mateer on 2020-05-22 17:08
```

Nice. Lots of talk on [https://www.reddit.com/r/do...](https://www.reddit.com/r/dotnet/) on MediatR and EF Core at the moment.

**Comment by giscus-migrator** [comment]
```header
Originally posted by Sunil Shetty on 2020-08-10 16:44
```

Thank you.. very good article

**Reply by giscus-migrator** [reply]
```header
Originally posted on 2020-08-16 17:30
```

Cheers mate

**Comment by giscus-migrator** [comment]
```header
Originally posted by Rasul Abbasi on 2020-11-25 22:55
```

Good article, But I have one important question, If we are using DDD the repository interface definitions must be in Domain layer otherwise you can't use Domain services in aggregates. Am I wrong?

**Reply by giscus-migrator** [reply]
```header
Originally posted on 2020-11-28 12:19
```

Hey Rasul. I'm no expert on DDD, but what you are saying makes sense. You could have the repository interfaces in the Domain layer, but keep the implementation in Infrastructure, so that the domain is still persistence ignorant.

**Comment by giscus-migrator** [comment]
```header
Originally posted by Rahul Agarawal on 2021-06-19 17:32
```

thanks for great article !! I have one question though - benefits you called out in context of clean architecture can be achieved in layered architecture with DI implementation, isn't ?

**Reply by giscus-migrator** [reply]
```header
Originally posted on 2021-06-19 17:54
```

Thanks for the feedback! Yes you can achieve some of the benefits if you use the layered approach and DI. It would really depend on which layers you have in your application and the depencies between them.

You will be missing out on some advantages of you dont use MediatR though.

**Comment by giscus-migrator** [comment]
```header
Originally posted by Şükrü Berkay on 2021-12-22 08:37
```

a clean article for clean architecture. 👏

**Reply by giscus-migrator** [reply]
```header
Originally posted on 2021-12-22 12:37
```

Cheers for the feedback!

**Comment by giscus-migrator** [comment]
```header
Originally posted by Rishi Patel on 2022-01-19 09:33
```

Nice one :)

**Comment by Jyuart** [comment]
Hello. Thank you for the article. Can you please elaborate on this point: "The Domain Layer could be included in the Application Layer, but if you are using an ORM like entity framework, the Infrastructure Layer will need to reference the domain models, in which case it's better to split out into a separate layer." 
Why does the use of ORM influences this decision?

**Reply by danielmackay** [reply]
Hi @Jyuart.  My view on this has changed slightly since writing the article.  Some CA solutions have a 'Core' Layer, which is a combination of both Domain and Application.  I think that is fine for trivial applications, but my preference is to keep Domain and Application in separate layers.  The benefit there is that we could share our Domain with multiple applications which is not possible when these are combined.

You are right, the use of an ORM or any other persistence technique should not influence where our Domain is stored.  One major point of a CA architecture is that the Domain should be persistence ignorant.

Thanks for the comment.

**Comment by arvin-yorro-prospa** [comment]
Hi Dan, thank you for this article.  I have a question regarding the Application Layer services.

> The Application Layer ONLY references the Domain Layer. It knows nothing of databases, web services, etc

I understand that it should reference outside layers or concrete implementations of the persistence layers.

My question is. Should it reference other "application layer services" as well, if no then why?  I'm in favor of NO but I'm looking for practical reasons for this. Thanks.

**Reply by danielmackay** [reply]
> I understand that it should reference outside layers or concrete implementations of the persistence layers.

If we're talking about the Application Layer, I think you mean "it should NOT" reference concrete implementations of the persistence layers.

> My question is. Should it reference other "application layer services" as well, if no then why? I'm in favor of NO but I'm looking for practical reasons for this. Thanks.

I'm not sure if I understand here.  Are you asking if Services in the Application Layer, can refernces other Services in the Application Layer?

Please correct me if I haven't understood you properly. 🙂

**Reply by arvin-yorro-prospa** [reply]
> If we're talking about the Application Layer, I think you mean "it should NOT" reference concrete implementations of the persistence layers.

Yes. Sorry that was a typo but thanks for catching that.

> Are you asking if Services in the Application Layer, can refernces other Services in the Application Layer?

Yes. Services in the Application Layer referencing other Services in the Application Layer (as dependencies). What do you think in your opinion? This was not specifically mentioned in your article.

I'm in favor of NO myself and I'm doing a research on why not.

Thanks Again.

**Reply by danielmackay** [reply]
> Yes. Services in the Application Layer referencing other Services in the Application Layer (as dependencies). What do you think in your opinion? This was not specifically mentioned in your article.
> 
> I'm in favor of NO myself and I'm doing a research on why not.

Generally I like to use the CQRS pattern in my Application layer.  This results in having a separate 'handler' for each command or query.  Each handle is also responsible for the transaction for each request (i.e. either everything or nothing gets saved).  Due to this transaction, it gets messy if handlers try to reference each other.  Therefore, I have a rule that they can't.  

If you're using services in the same capacity, then I would suggest the same rules apply.
